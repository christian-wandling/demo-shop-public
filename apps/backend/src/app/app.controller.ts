import { Controller, Get } from '@nestjs/common';
import { Public } from 'nest-keycloak-connect';

@Controller('app')
export class AppController {
  @Public()
  @Get('/debug-sentry')
  getError() {
    throw new Error('My first Sentry error!');
  }
}
