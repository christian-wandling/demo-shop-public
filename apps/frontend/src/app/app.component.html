@if (initialized()) {
<lib-cart></lib-cart>
<lib-navigation></lib-navigation>
<router-outlet></router-outlet>
}
