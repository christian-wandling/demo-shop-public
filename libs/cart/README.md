# `@demo-shop/cart`

This library handles all aspects of the shopping cart feature including product selection, cart management, and the checkout process.

## Features

- **Shopping Cart Management**

  - Add/remove items
  - Update quantities

- **Checkout Process**

  - Address management
  - Order review
  - Order confirmation

- **State Management**
  - Signal based state
  - Synchronization with backend

## API Reference

For a complete API reference, see the [Compodoc documentation](https://christian-wandling.github.io/demo-shop-public/compodoc/frontend).

## Running Unit Tests

Run `nx test cart` to execute the unit tests.
