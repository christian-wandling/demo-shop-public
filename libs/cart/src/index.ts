export * from './lib/cart.routes';
export * from './lib/cart.facade';
export * from './lib/components/cart/cart.component';
export * from './lib/components/checkout/checkout.component';
export * from './lib/components/cart-icon/cart-icon.component';
