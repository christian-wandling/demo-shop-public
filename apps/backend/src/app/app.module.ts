import { Module } from '@nestjs/common';
import { ProductModule } from './product/product.module';
import { ConfigModule } from '@nestjs/config';
import { UserModule } from './user/user.module';
import { OrderModule } from './order/order.module';
import { CommonModule } from './common/common.module';
import { ShoppingSessionModule } from './shopping-session/shopping-session.module';
import { CartItemModule } from './cart-item/cart-item.module';
import { JwtModule } from '@nestjs/jwt';
import {
  AuthGuard,
  KeycloakConnectModule,
  PolicyEnforcementMode,
  RoleGuard,
  TokenValidation,
} from 'nest-keycloak-connect';
import { APP_FILTER, APP_GUARD } from '@nestjs/core';
import { ThrottlerGuard, ThrottlerModule } from '@nestjs/throttler';
import { SentryGlobalFilter, SentryModule } from '@sentry/nestjs/setup';
import { AppController } from './app.controller';

@Module({
  imports: [
    SentryModule.forRoot(),
    ConfigModule.forRoot(),
    JwtModule.register({
      global: true,
    }),
    ThrottlerModule.forRoot([
      {
        ttl: 60000,
        limit: 10,
      },
    ]),
    KeycloakConnectModule.register({
      authServerUrl: process.env.KEYCLOAK_URL,
      realm: process.env.KEYCLOAK_REALM,
      clientId: process.env.KEYCLOAK_CLIENT_API,
      secret: '',
      tokenValidation: TokenValidation.OFFLINE,
      policyEnforcement: PolicyEnforcementMode.PERMISSIVE,
      realmPublicKey: process.env.KEYCLOAK_REALM_PUBLIC_KEY,
      bearerOnly: true,
    }),
    CommonModule,
    ProductModule,
    UserModule,
    CartItemModule,
    OrderModule,
    ShoppingSessionModule,
  ],
  providers: [
    {
      provide: APP_FILTER,
      useClass: SentryGlobalFilter,
    },
    {
      provide: APP_GUARD,
      useClass: AuthGuard,
    },
    {
      provide: APP_GUARD,
      useClass: RoleGuard,
    },
    {
      provide: APP_GUARD,
      useClass: ThrottlerGuard,
    },
  ],
  controllers: [AppController],
})
export class AppModule {}
