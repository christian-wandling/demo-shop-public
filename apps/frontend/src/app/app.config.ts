import { APP_INITIALIZER, ApplicationConfig, ErrorHandler, isDevMode, provideZoneChangeDetection } from '@angular/core';
import { provideRouter, Router, RouteReuseStrategy, withRouterConfig } from '@angular/router';
import { appRoutes } from './app.routes';
import { provideStoreDevtools } from '@ngrx/store-devtools';
import { provideHttpClient, withFetch, withInterceptors } from '@angular/common/http';
import { provideAnimationsAsync } from '@angular/platform-browser/animations/async';
import { provideRouterStore, routerReducer } from '@ngrx/router-store';
import { provideStore } from '@ngrx/store';
import { provideApi, withConfiguration } from './providers/provide-api';
import { authInterceptor, provideAuth } from '@demo-shop/auth';
import { environment } from '../environments/environment';
import * as Sentry from '@sentry/angular';
import { NoReuseStrategy } from './strategies/no-reuse-strategy';
import { provideMonitoring } from '@demo-shop/monitoring';

export const appConfig: ApplicationConfig = {
  providers: [
    /** Performance optimization for change detection */
    provideZoneChangeDetection({ eventCoalescing: true }),

    /** HTTP client with fetch API and authentication interceptor */
    provideHttpClient(withFetch(), withInterceptors([authInterceptor])),

    /** Router configuration with reload behavior on same URL navigation */
    provideRouter(appRoutes, withRouterConfig({ onSameUrlNavigation: 'reload' })),

    /** Async loading of animation capabilities */
    provideAnimationsAsync(),

    /** Redux DevTools configuration for state debugging */
    provideStoreDevtools({ maxAge: 25, logOnly: !isDevMode() }),

    /** Router state tracking integration */
    provideRouterStore(),

    /** NgRx store setup with router reducer */
    provideStore({
      router: routerReducer,
    }),

    /** Custom route reuse strategy to prevent component reuse */
    {
      provide: RouteReuseStrategy,
      useClass: NoReuseStrategy,
    },

    /** Sentry error handler integration */
    {
      provide: ErrorHandler,
      useValue: Sentry.createErrorHandler(),
    },

    /** Sentry performance tracing service */
    {
      provide: Sentry.TraceService,
      deps: [Router],
    },

    /**  Initialize Sentry tracing on application startup */
    {
      provide: APP_INITIALIZER,
      useFactory: () => () => undefined,
      deps: [Sentry.TraceService],
      multi: true,
    },

    /** API client configuration with basePath: '' */
    provideApi(withConfiguration({ basePath: '' })),

    /** Provide Keycloak configuration to Auth lib */
    provideAuth({ keycloak: environment.keycloak }),

    /** Provide Sentry configuration to monitoring lib */
    provideMonitoring({ sentry: environment.sentry }),
  ],
};
