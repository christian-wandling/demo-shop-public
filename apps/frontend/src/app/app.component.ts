import { ChangeDetectionStrategy, Component, inject } from '@angular/core';
import { RouterModule } from '@angular/router';
import { NavigationComponent } from '@demo-shop/navigation';
import { CartComponent } from '@demo-shop/cart';
import { AppStore } from './+state/app.store';

@Component({
  standalone: true,
  imports: [RouterModule, NavigationComponent, CartComponent],
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrl: './app.component.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class AppComponent {
  readonly #appStore = inject(AppStore);
  readonly initialized = this.#appStore.initialized;
}
