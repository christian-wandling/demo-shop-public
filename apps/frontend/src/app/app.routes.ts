import { Route } from '@angular/router';
import { CheckoutComponent } from '@demo-shop/cart';

export const appRoutes: Route[] = [
  {
    path: '',
    redirectTo: 'products',
    pathMatch: 'full',
  },
  {
    path: 'checkout',
    component: CheckoutComponent,
    data: { pageTitle: 'Checkout', showInMenu: false },
  },
  {
    path: 'orders',
    loadChildren: () => import('@demo-shop/order').then(m => m.orderRoutes),
    data: { pageTitle: 'Orders', showInMenu: true },
  },
  {
    path: 'products',
    loadChildren: () => import('@demo-shop/product').then(m => m.productRoutes),
    data: { pageTitle: 'Products', showInMenu: true },
  },
];
