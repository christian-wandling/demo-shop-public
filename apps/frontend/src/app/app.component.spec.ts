import { ComponentFixture, TestBed } from '@angular/core/testing';
import { AppComponent } from './app.component';
import { AppStore } from './+state/app.store';
import { Component, signal } from '@angular/core';
import { provideRouter } from '@angular/router';
import { NavigationComponent } from '@demo-shop/navigation';
import { CartComponent } from '@demo-shop/cart';
import { By } from '@angular/platform-browser';

describe('AppComponent', () => {
  let component: AppComponent;
  let fixture: ComponentFixture<AppComponent>;

  const appStoreStub = {
    initialized: signal(false),
  };

  // eslint-disable-next-line @angular-eslint/component-selector
  @Component({ standalone: true, selector: 'lib-navigation', template: '' })
  class NavigationStubComponent {}

  // eslint-disable-next-line @angular-eslint/component-selector
  @Component({ standalone: true, selector: 'lib-cart', template: '' })
  class CartStubComponent {}

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      providers: [{ provide: AppStore, useValue: appStoreStub }, provideRouter([])],
      imports: [AppComponent],
    })
      .overrideComponent(AppComponent, {
        remove: { imports: [NavigationComponent, CartComponent] },
        add: { imports: [NavigationStubComponent, CartStubComponent] },
      })
      .compileComponents();

    fixture = TestBed.createComponent(AppComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should not display navigation and cart before initialization', () => {
    expect(component.initialized()).toBe(false);
    expect(fixture).toMatchSnapshot();
  });

  it('should display navigation and cart before initialization', () => {
    appStoreStub.initialized.set(true);
    fixture.detectChanges();

    expect(component.initialized()).toBe(true);
    expect(fixture.debugElement.query(By.css('lib-navigation'))).toBeTruthy();
    expect(fixture.debugElement.query(By.css('lib-cart'))).toBeTruthy();
    expect(fixture).toMatchSnapshot();
  });
});
