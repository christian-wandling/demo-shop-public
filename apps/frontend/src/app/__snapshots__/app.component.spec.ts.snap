// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should display navigation and cart before initialization 1`] = `
<app-root
  initialized={[Function Function]}
>
  <lib-cart /><lib-navigation /><router-outlet />
</app-root>
`;

exports[`AppComponent should not display navigation and cart before initialization 1`] = `
<app-root
  initialized={[Function Function]}
>
</app-root>
`;
